janus_server/src/hpke.rs

```diff
@@ -0,0 +1,379 @@
+            )),
+        }
+    }
+}
+
+// This function exists separately from struct HpkeSender to abstract away its
+// generic parameters
+fn seal<Encrypt: Aead, Derive: Kdf, Encapsulate: Kem>(
+    recipient_config: &HpkeConfig,
+    application_info: HpkeApplicationInfo,
+    plaintext: &[u8],
+    associated_data: &[u8],
+) -> Result<HpkeCiphertext, Error> {
+    let mut rng = thread_rng();
+
+    // Deserialize recipient pub into the appropriate PublicKey type for the
+    // KEM
+    let recipient_public_key = Encapsulate::PublicKey::from_bytes(&recipient_config.public_key.0)?;
+
+    let (encapsulated_context, mut context) = setup_sender::<Encrypt, Derive, Encapsulate, _>(
+        &OpModeS::Base,
+        &recipient_public_key,
+        &application_info.0,
+        &mut rng,
+    )?;
+
+    Ok(HpkeCiphertext {
+        config_id: recipient_config.id,
+        encapsulated_context: encapsulated_context.to_bytes().to_vec(),
+        payload: context.seal(plaintext, associated_data)?,
+    })
+}
+
+/// An HPKE recipient that decrypts messages encrypted to the public key in
+/// `recipient_config`, using the AEAD, key derivation and key encapsulation
+/// mechanisms specified in `recipient_config`, and using `label`, `sender_role`
+/// and `recipient_role` to derive application info.
+//
+// This type only exists separately from Recipient so that we can have a type
+// that doesn't "leak" the generic type parameters into the caller.
+struct HpkeRecipient {
+    task_id: TaskId,
+    config: HpkeConfig,
+    label: Label,
+    sender_role: Role,
+    recipient_role: Role,
+    recipient_private_key: HpkePrivateKey,
+}
+
+impl HpkeRecipient {
+    /// Generate a new X25519HkdfSha256 keypair and construct an HPKE recipient
+    /// using the private key, with KEM = X25519HkdfSha256, KDF = HkdfSha512 and
+    /// AEAD = ChaCha20Poly1305, and the specified label, and roles.
+    fn generate(label: Label, sender_role: Role, recipient_role: Role) -> Self {
+        let mut rng = thread_rng();
+
+        let task_id = TaskId::random();
+
+        let (private_key, public_key) = X25519HkdfSha256::gen_keypair(&mut rng);
+
+        let config = HpkeConfig {
+            id: HpkeConfigId(0),
+            kem_id: HpkeKemId::X25519HkdfSha256,
+            kdf_id: HpkeKdfId::HkdfSha512,
+            aead_id: HpkeAeadId::ChaCha20Poly1305,
+            public_key: HpkePublicKey(public_key.to_bytes().as_slice().to_vec()),
+        };
+
+        Self {
+            task_id,
+            config,
+            label,
+            sender_role,
+            recipient_role,
+            recipient_private_key: HpkePrivateKey(private_key.to_bytes().as_slice().to_vec()),
+        }
+    }
+
+    /// Decrypt `ciphertext` and return the plaintext.
+    ///
+    /// In PPM, an HPKE context can only be used once (we have no means of
+    /// ensuring that sender and recipient "increment" nonces in lockstep), so
+    /// this method consumes self.
```

Comment:
In the bare function `open()` below, we create a context, use it once, and drop it, which should be sufficient to ensure we only use an HPKE context once (with its initial nonce only). As `HpkeRecipient` contains an `HpkeConfig`, private key, and AAD-related fields, it should be fine for it to be reused, and moreover we may want to do so when handling a series of aggregate requests, rather than construct this struct repeatedly, copying both sides of the keypair each time.

Reply:
Done!